subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,334 @@
+
+
+----
+// Without kotlin-asignment
+tasks.jar.configure {
+    // You need to call .set()
+    archiveBaseName.set("mylib")
+}
+
+// With kotlin-asignment
+tasks.jar.configure {
+    // No need to call .set()
+    archiveBaseName = "mylib"
+}
+----
+
+[[sec:api_version]]
+=== `apiVersion` compiler option
+
+The https://kotlinlang.org/docs/compiler-reference.html#api-version-version[Kotlin compiler apiVersion option] enforces that the generated bytecode only uses declaration from the given version. This is important to stay compatible with the Kotlin version embedded by Gradle.
+
+Note: while this enforces your code is compatible with a given version of Kotlin stdlib and Gradle, it does not check dependencies, so it is important to check your dependencies are also compatible.
+
+[[sec:sam_conversion_class]]
+=== `-Xsam-conversions=class` compiler option
+
+`-Xsam-conversions=class` sets up the implementation strategy for SAM (single abstract method) conversion to always generate anonymous classes, instead of using the `invokedynamic` JVM instruction. This is to make sure the lambdas are serializable and provide better support for configuration cache and incremental build. This is particularly important for lambdas used from `doLast`:
+See https://github.com/gradle/gradle/issues/17052[Issue #17052] for more information.
+
+[source,kotlin]
+----
+// Without `-Xsam-conversions=class` task will never be up-to-date
+task.doLast {
+    // Something
+}
+----
+
+
+[[sec:plugins_for_your_plugins]]
+== Different development plugins
+
+In order to configure all the above options and more, Gradle offers several plugins:
+
+* `"java-gradle-plugin"`
+* `kotlin("jvm")` (id `"org.jetbrains.kotlin.jvm"`)
+* `embedded-kotlin` (id `"org.gradle.kotlin.embedded-kotlin"`))
+* `kotlin-dsl` (id `"org.gradle.kotlin.kotlin-dsl"`)
+
+Usually you would choose one of those combinations:
+
+* `"java-gradle-plugin"` and `kotlin("jvm")`
+* `"java-gradle-plugin"` and `embedded-kotlin`
+* `kotlin-dsl`
+
+Let's see what each plugin is doing.
+
+[[sec:java_gradle_plugin]]
+=== The `"java-gradle-plugin"` plugin (built-in)
+
+`"java-gradle-plugin"` amongst other things:
+
+* applies the <<java_library_plugin.adoc#java_library_plugin,Java Library (`java-library`)>> plugin
+* adds the `gradlePlugin {}` extension and generates the resource files matching your plugin id with its implementation class
+* adds the `gradleApi()` dependency to the `api` configuration
+* performs validation of plugin metadata during `jar` task execution.
+* adds the `gradleTestKit()` dependency to the `testImplementation` configuration
+
+You always want this plugin to develop custom Gradle plugins (although other plugins might apply it for you). Refer to the full <<java_gradle_plugin.adoc, documentation>> for more information.
+
+[[sec:kotlin_jvm]]
+=== The `kotlin("jvm")` plugin
```

Comment:
This can become very confusing for somebody who doesn't know that the "kotlin(String)" function here is just a helper for building plugin IDs. Better use raw plugin IDs instead.

Reply:
Yea agreed. See also [this comment](https://github.com/gradle/gradle/pull/24286/files#r1135314524). I'd love to remove `kotlin("jvm")`.

I'm updating this PR to use the plugin id primarily for Kotlin Gradle Plugin and keep `kotlin-embedded` and `kotlin-dsl` as is because there is just too much brain muscle